File: ruby/tm/Tape.rb
#!/usr/bin/env ruby

class Tape < Struct.new(:left, :middle, :right, :blank)
  def inspect
    "#<Tape #{left.join}(#{middle})#{right.join}> "
  end

  def to_s
    inspect
  end

  def write(character)
    Tape.new(left, character, right, blank)
  end

  def move_head_left
    Tape.new(left[0..-2], left.last || blank, [middle] + right)
  end

  def move_head_right
    Tape.new(left + [middle], right.first || blank, right.drop(1), blank)
  end
end

tape = Tape.new(['1', '0', '1'], '1', [], '_')
puts tape
puts tape.move_head_left()
puts tape.write('0')

File: javascript/tm/tape.js
'use strict';

class Tape {
  constructor({ left, middle, right, blank }) {
    this.left = left;
    this.middle = middle;
    this.right = right;
    this.blank = blank;
  }

  write(character) {
    return new Tape({ ...this, middle: character });
  }

  moveHeadLeft() {
    const leftLast = this.left.slice(-1)

    return new Tape({
      left: this.left.slice(0, this.left.length -1),
      middle: leftLast[0] || this.blank,
      right: leftLast.concat(this.right.slice()),
      blank: this.blank
    });
  }

  moveHeadRight() {
    const right = this.right.slice();

    return new Tape({
      left: this.left.slice().concat([this.middle]),
      middle: right.shift() || this.blank,
      right,
      blank: this.blank
    });
  }
}

module.exports = Tape
